src/Core/Name/Scoped.idr

```diff
@@ -0,0 +1,264 @@
+Scoped : Type
+Scoped = Scope -> Type
+
+------------------------------------------------------------------------
+-- Semi-decidable equality
+
+export
+scopeEq : (xs, ys : Scope) -> Maybe (xs = ys)
+scopeEq [] [] = Just Refl
+scopeEq (x :: xs) (y :: ys)
+    = do Refl <- nameEq x y
+         Refl <- scopeEq xs ys
+         Just Refl
+scopeEq _ _ = Nothing
+
+------------------------------------------------------------------------
+-- Generate a fresh name (for a given scope)
+
+export
+mkFresh : Scope -> Name -> Name
+mkFresh vs n
+  = if n `elem` vs
+    then assert_total $ mkFresh vs (next n)
+    else n
+
+
+------------------------------------------------------------------------
+-- Compatible variables
+
+public export
+data CompatibleVars : (xs, ys : List a) -> Type where
+   Pre : CompatibleVars xs xs
+   Ext : CompatibleVars xs ys -> CompatibleVars (n :: xs) (m :: ys)
+
+export
+invertExt : CompatibleVars (n :: xs) (m :: ys) -> CompatibleVars xs ys
+invertExt Pre = Pre
+invertExt (Ext p) = p
+
+export
+extendCompats : (args : List a) ->
+                CompatibleVars xs ys ->
+                CompatibleVars (args ++ xs) (args ++ ys)
+extendCompats args Pre = Pre
+extendCompats args prf = go args prf where
+
+  go : (args : List a) ->
+       CompatibleVars xs ys ->
+       CompatibleVars (args ++ xs) (args ++ ys)
+  go [] prf = prf
+  go (x :: xs) prf = Ext (go xs prf)
+
+export
+decCompatibleVars : (xs, ys : List a) -> Dec (CompatibleVars xs ys)
+decCompatibleVars [] [] = Yes Pre
+decCompatibleVars [] (x :: xs) = No (\case p impossible)
+decCompatibleVars (x :: xs) [] = No (\case p impossible)
+decCompatibleVars (x :: xs) (y :: ys) = case decCompatibleVars xs ys of
+  Yes prf => Yes (Ext prf)
+  No nprf => No (nprf . invertExt)
+
+export
+areCompatibleVars : (xs, ys : List a) ->
+                    Maybe (CompatibleVars xs ys)
+areCompatibleVars [] [] = pure Pre
+areCompatibleVars (x :: xs) (y :: ys)
+    = do compat <- areCompatibleVars xs ys
+         pure (Ext compat)
+areCompatibleVars _ _ = Nothing
+
+------------------------------------------------------------------------
+-- Thinnings
+
+public export
+data Thin : List a -> List a -> Type where
+  Refl : Thin xs xs
+  Drop : Thin xs ys -> Thin xs (y :: ys)
+  Keep : Thin xs ys -> Thin (x :: xs) (x :: ys)
+
+export
+none : {xs : List a} -> Thin [] xs
+none {xs = []} = Refl
+none {xs = _ :: _} = Drop none
+
+||| Smart constructor. We should use this to maximise the length
+||| of the Refl segment thus getting more short-circuiting behaviours
+export
+keep : Thin xs ys -> Thin (x :: xs) (x :: ys)
+keep Refl = Refl
+keep p = Keep p
+
+export
+keeps : (args : List a) -> Thin xs ys -> Thin (args ++ xs) (args ++ ys)
+keeps [] th = th
+keeps (x :: xs) th = keep (keeps xs th)
+
+namespace Thin
+  -- At runtime, Thin's `Refl` does not carry any additional
+  -- information. So this is safe!
+  export
+  embed : Thin xs ys -> Thin (xs ++ outer) (ys ++ outer)
+  embed = believe_me
```

Comment:
The following definition:
```haskell

embed : Thin xs ys -> Thin (xs ++ outer) (ys ++ outer)
embed Refl = Refl
embed (Drop x) = Drop (embed x)
embed (Keep x) = Keep (embed x)
```
compiles to the identity function and avoids using `believe_me`.
On the other hand using `believe_me` is less fragile in case that optimisation is removed. (The optimisation here is the same one used to optimise `toInteger` and `fromInteger` for arbitrary `Nat` shaped types)

Reply:
Yep. It's not even used in the codebase AFAICT so I may as well just drop it.